testcases/unicode-strings.http

```diff
@@ -0,0 +1,9 @@
+POST / HTTP/1.1
+Host: handler.example.com
+Content-Length: 0
+ce-specversion: 0.4-wip
+ce-id: %22%F0%9F%98%89%03%3c.is%2BFine%E7%81%AB%22
```

Comment:
after looking at the http spec, I'm not sure this is correct. While someone could URL-encode the header, unless some spec tells both sides about it, I don't think it's supposed to be. By that I mean, I couldn't find anything in the http spec to indicate people are supposed to URL encode things - and therefore, there's no mechanism for a receiver to know if they're suppose to try to decode it or not.

If these are printable chars, then no encoding would be done. If they're non-printable chars then the spec bans them as part of our string definition.

Reply:
https://tools.ietf.org/html/rfc7230#section-3.2.4

>    Historically, HTTP has allowed field content with text in the
>    ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
>    through use of [RFC2047] encoding.  In practice, most HTTP header
>    field values use only a subset of the US-ASCII charset [USASCII].
>    Newly defined header fields SHOULD limit their field values to
>    US-ASCII octets.  A recipient SHOULD treat other octets in field
>    content (obs-text) as opaque data.

Note that this conflicts with our definition of String. Maybe we should add some text on how to deal with unicode characters in the HTTP binary encoding (and maybe other binary encodings, depending on their allowed charset)?

Reply:
Following up on this, it turns out that someone back in April of 2018 spelled out how this is supposed to work, and it is by URL-encoding the strings:

https://github.com/cloudevents/spec/blob/master/http-transport-binding.md#3132-http-header-values

> Non-printable ASCII characters and non-ASCII characters MUST first be encoded according to UTF-8, and then each octet of the corresponding UTF-8 sequence MUST be percent-encoded to be represented as HTTP header characters, in compliance with RFC7230, sections 3, 3.2, 3.2.6. The rules for encoding of the percent character ('%') apply as defined in RFC 3986 Section 2.4..